src/nlv_object.h

```diff
@@ -5,182 +5,184 @@
+      Worker::OnFinishedHandler operator()(const char* val) {
+        std::string str = val;
+        return [=](Worker* worker) {
+          Nan::HandleScope scope;
+          v8::Local<v8::Value> argv[] = { Nan::Null(), Nan::New(str).ToLocalChecked() };
+          worker->Call(2, argv);
+        };
+      }
+    };
+    
+    static inline bool virHasFailed(int val) {
+      return val < 0;
+    }
+    
+    static inline bool virHasFailed(const void* ptr) {
+      return ptr == nullptr;
+    }
 
-  typedef typename std::remove_pointer<HandleType>::type HandleValue;
+    template<typename ReturnHandler = MethodNoReturnHandler, typename Z, typename... Args>
+    static void RunMethod(const Nan::FunctionCallbackInfo<v8::Value>& info, Z virFunction, Args... args) {
+      Nan::HandleScope scope;
+      virDomainPtr handle = ParentClass::Unwrap(info.This())->virHandle();
+      auto virCaller = std::bind(virFunction, handle, args...);
+      NLV::Worker::RunAsync(info, [=] (NLV::Worker::SetOnFinishedHandler onFinished) {
+        auto retVal = virCaller();
+        if (virHasFailed(retVal)) {
+          return virSaveLastError();
+        }
+        ReturnHandler returnHandler;
+        return onFinished(returnHandler(retVal));
+      });
+    }
 
-  NLVObject(HandleType handle) : handle_(handle, CleanupHandler::cleanup), parentReference_(NULL) {}
-  ~NLVObject() {
-    // calling virtual ClearHandle() will break if overridden by subclasses
-    ClearHandle();
-    if (parentReference_ != NULL) {
-      parentReference_->SetInvalid();
+    static v8::Local<v8::Object> NewInstance(handle_type handle) {
+      Nan::EscapableHandleScope scope;
+      Local<Function> ctor = Nan::New<Function>(ParentClass::constructor);
+      Local<Object> object = Nan::NewInstance(ctor).ToLocalChecked();
+      ParentClass *class_instance = new ParentClass(handle);
+      class_instance->Wrap(object);
+      return scope.Escape(object);
     }
-  }
 
-  static v8::Local<v8::Object> NewInstance(handle_type handle) {
-    Nan::EscapableHandleScope scope;
-    Local<Function> ctor = Nan::New<Function>(ParentClass::constructor);
-    Local<Object> object = Nan::NewInstance(ctor).ToLocalChecked();
-    ParentClass *class_instance = new ParentClass(handle);
-    class_instance->Wrap(object);
-    return scope.Escape(object);
-  }
+    static bool IsInstanceOf(v8::Local<v8::Object> val) {
+      Nan::HandleScope scope;
+      return Nan::New(ParentClass::constructor_template)->HasInstance(val);
+    }
 
-  static bool IsInstanceOf(v8::Local<v8::Object> val) {
-    Nan::HandleScope scope;
-    return Nan::New(ParentClass::constructor_template)->HasInstance(val);
-  }
+    const HandleType virHandle() const {
+      return handle_.get();
+     }
 
-  const HandleType virHandle() const {
-    return handle_.get();
-   }
 
+    NAN_INLINE static ParentClass* Unwrap(v8::Local<v8::Object> val) {
+      if(!ParentClass::IsInstanceOf(val)) {
+        char error[128];
```

Comment:
It's from a previous merge, I just had to move this code around. 

It can probably be rewritten to `Nan::ThrowTypeError(std::string("Expecting object to be ") + ParentClass::ClassName());`